### PR TITLE
feature(pu/zt): add 2048 env and Stochastic MuZero

### PR DESCRIPTION
**Pull request by puyuan1996**
> - add stochastic muzero ptree and ctree, policy, model
> - add 2048 env and unittest
> - add 2048 bot
> - add 2048 related configs

- The benchmark results in the 2048 environment:

- The render video to illustrate the performance of Random action, Bot action, MuZero, and Stochastic MuZero agents in the 2048 environment.

    - Random action (~1000):
game_2048_random

    - Bot action (~30000-70000):
game_2048_bot_small

    - MuZero sim=100 (~27956):
game_2048_muzero_ns100_s0

    - Stochastic MuZero sim=100 (~37356）:
game_2048_stochastic_muzero_ns100_s0_small

    - Where the term `sim` represents the number of simulations used for evaluation in MCTS, and the score within brackets corresponds to the score procured by each respective agent.

#41

### CODE REVIEW

File: zoo/game_2048/config/stochastic_muzero_2048_config.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
+from easydict import EasyDict
+import sys
+sys.path.append('/mnt/nfs/puyuan/LightZero/zoo/game_2048')
```
Comment (PaParaZz1) [inline zoo/game_2048/config/stochastic_muzero_2048_config.py:3]: don't add personal info

File: zoo/game_2048/config/stochastic_muzero_2048_eval_config.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+# According to the model you want to evaluate, import the corresponding config.
```
Comment (PaParaZz1) [inline zoo/game_2048/config/stochastic_muzero_2048_eval_config.py:1]: move to entry directory

File: zoo/game_2048/config/stochastic_muzero_2048_config.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            # NOTE: whether to use the self_supervised_learning_loss. default is False
+            self_supervised_learning_loss=True,  # default is False
+            discrete_action_encoding_type='one_hot',
+            norm_type='BN',
+        ),
+        use_ture_chance_label_in_chance_encoder=use_ture_chance_label_in_chance_encoder,
+        mcts_ctree=True,
+        gumbel_algo=False,
+        cuda=True,
+        env_type='not_board_games',
```
Comment (PaParaZz1) [inline zoo/game_2048/config/stochastic_muzero_2048_config.py:69]: simplify config fields

File: zoo/game_2048/config/muzero_2048_config.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    env=dict(
+        type='game_2048',
+        import_names=['zoo.game_2048.envs.game_2048_env'],
+    ),
+    env_manager=dict(type='subprocess'),
+    policy=dict(
+        type='muzero',
+        import_names=['lzero.policy.muzero'],
+    ),
+    collector=dict(
```
Comment (PaParaZz1) [inline zoo/game_2048/config/muzero_2048_config.py:100]: remove this field, collector is created by policy type in entry

File: lzero/policy/muzero.py [hunk 5]
```diff
@@ -221,7 +223,11 @@ def _init_learn(self) -> None:
             self._optimizer = optim.Adam(
                 self._model.parameters(), lr=self._cfg.learning_rate, weight_decay=self._cfg.weight_decay
             )
-        elif self._cfg.optim_type == 'AdamW':
+        elif self._cfg.optim_type == 'AdamW_official':
```
Comment (PaParaZz1) [inline lzero/policy/muzero.py:226]: why modify this

File: lzero/mcts/buffer/game_segment.py [hunk 6]
```diff
@@ -128,7 +131,8 @@ def append(
             obs: np.ndarray,
             reward: np.ndarray,
             action_mask: np.ndarray = None,
-            to_play: int = -1
+            to_play: int = -1,
+            chance: np.ndarray = 0,
```
Comment (PaParaZz1) [inline lzero/mcts/buffer/game_segment.py:135]: why is the default value a python scalar

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: zoo/game_2048/config/stochastic_muzero_2048_config.py (hunk 1 above)
Comment (puyuan1996) [reply zoo/game_2048/config/stochastic_muzero_2048_config.py:4]: ok

File: zoo/game_2048/config/stochastic_muzero_2048_eval_config.py (hunk 2 above)
Comment (puyuan1996) [reply zoo/game_2048/config/stochastic_muzero_2048_eval_config.py:1]: ok

File: zoo/game_2048/config/stochastic_muzero_2048_config.py (hunk 3 above)
Comment (puyuan1996) [reply zoo/game_2048/config/stochastic_muzero_2048_config.py:73]: ok

File: lzero/policy/muzero.py (hunk 5 above)
Comment (puyuan1996) [reply lzero/policy/muzero.py:226]: wrongly modified, fixed now

File: lzero/mcts/buffer/game_segment.py (hunk 6 above)
Comment (puyuan1996) [reply lzero/mcts/buffer/game_segment.py:135]: fixed now

File: zoo/game_2048/config/muzero_2048_config.py (hunk 4 above)
Comment (puyuan1996) [reply zoo/game_2048/config/muzero_2048_config.py:107]: polished now

File: lzero/policy/stochastic_muzero.py [hunk 7]
```diff
@@ -0,0 +1,815 @@
...
+            policy_loss += cross_entropy_loss(policy_logits, target_policy[:, step_i + 1])
+
+            # TODO(pu):
+            if self._cfg.use_ture_chance_label_in_chance_encoder:
+                afterstate_policy_loss += cross_entropy_loss(afterstate_policy_logits, true_chance_one_hot.detach())
+                # The chance encoder is not used in the mcts, so we don't need to calculate the commitment loss.
+                commitment_loss += torch.nn.MSELoss()(chance_encoding, true_chance_one_hot.float().detach())
+            else:
+                afterstate_policy_loss += cross_entropy_loss(afterstate_policy_logits, chance_one_hot.detach())
+                # TODO(pu): whether to detach the chance_one_hot in the commitment loss.
```
Comment (puyuan1996) [inline lzero/policy/stochastic_muzero.py:454]: 这个地方需要确认一下

File: lzero/policy/muzero.py [hunk 8]
```diff
@@ -208,7 +210,7 @@ def _init_learn(self) -> None:
         Overview:
             Learn mode init method. Called by ``self.__init__``. Initialize the learn model, optimizer and MCTS utils.
         """
-        assert self._cfg.optim_type in ['SGD', 'Adam', 'AdamW'], self._cfg.optim_type
+        assert self._cfg.optim_type in ['SGD', 'Adam', 'AdamW_official', 'AdamW_nanoGPT'], self._cfg.optim_type
```
Comment (PaParaZz1) [inline lzero/policy/muzero.py:213]: remove these two branches `'AdamW_official', 'AdamW_nanoGPT'`

File: lzero/policy/stochastic_muzero.py [hunk 9]
```diff
@@ -0,0 +1,828 @@
...
+    configure_optimizers
+from lzero.policy.utils import calculate_topk_accuracy, plot_topk_accuracy, visualize_avg_softmax, \
+    plot_argmax_distribution
+
+
+@POLICY_REGISTRY.register('stochastic_muzero')
+class StochasticMuZeroPolicy(Policy):
+    """
+    Overview:
+        The policy class for Stochastic MuZero.
```
Comment (PaParaZz1) [inline lzero/policy/stochastic_muzero.py:27]: add paper link

File: lzero/policy/stochastic_muzero.py [hunk 10]
```diff
@@ -0,0 +1,828 @@
...
+                - model_type (:obj:`str`): The model type used in this algorithm, which is registered in ModelRegistry.
+                - import_names (:obj:`List[str]`): The model class path list used in this algorithm.
+        .. note::
+            The user can define and use customized network model but must obey the same interface definition indicated \
+            by import_names path. For MuZero, ``lzero.model.muzero_model.MuZeroModel``
+        """
+        if self._cfg.model.model_type == "conv":
+            return 'StochasticMuZeroModel', ['lzero.model.stochastic_muzero_model']
+        elif self._cfg.model.model_type == "mlp":
+            return 'StochasticMuZeroModelMLP', ['lzero.model.stochastic_muzero_model_mlp']
```
Comment (PaParaZz1) [inline lzero/policy/stochastic_muzero.py:209]: add else branch to raise `KeyError`

File: lzero/policy/stochastic_muzero.py [hunk 11]
```diff
@@ -0,0 +1,828 @@
...
+            obs_batch = self.image_transforms.transform(obs_batch)
+            if self._cfg.model.self_supervised_learning_loss:
+                obs_target_batch = self.image_transforms.transform(obs_target_batch)
+
+        # shape: (batch_size, num_unroll_steps, action_dim)
+        # NOTE: .long(), in discrete action space.
+        action_batch = torch.from_numpy(action_batch).to(self._cfg.device).unsqueeze(-1).long()
+        data_list = [
+            mask_batch,
+            target_reward.astype('float64'),
```
Comment (PaParaZz1) [inline lzero/policy/stochastic_muzero.py:318]: why float64, all the float data shoule be float32

File: lzero/policy/utils.py [hunk 12]
```diff
@@ -1,18 +1,158 @@
...
+    plt.figure(figsize=(10, 8))
+    plt.bar(np.arange(len(avg_probabilities_np)), avg_probabilities_np)
+
+    plt.xlabel('Classes')
+    plt.ylabel('Average Probability')
+    plt.title('Average Softmax Probabilities Across the Minibatch')
+    plt.show()
+
+
+def calculate_topk_accuracy(logits, true_one_hot, top_k):
```
Comment (PaParaZz1) [inline lzero/policy/utils.py:40]: add unittest for these functions

File: zoo/game_2048/entry/rule_based_2048_config.py [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+
+import numpy as np
+from easydict import EasyDict
+from rich import print
+
+from zoo.game_2048.envs.game_2048_env import Game2048Env
+
+
+# Define rule-based search function
+def rule_based_search(grid: np.array, fast_search: bool = True) -> int:
```
Comment (PaParaZz1) [inline zoo/game_2048/entry/rule_based_2048_config.py:12]: separate these search functions to env directory

File: zoo/game_2048/envs/game_2048_env.py [hunk 14]
```diff
@@ -0,0 +1,687 @@
...
+from ding.utils import ENV_REGISTRY
+from easydict import EasyDict
+from gym import spaces
+from gym.utils import seeding
+
+
+@ENV_REGISTRY.register('game_2048')
+class Game2048Env(gym.Env):
+    config = dict(
+        env_name="game_2048",
```
Comment (PaParaZz1) [inline zoo/game_2048/envs/game_2048_env.py:23]: add comments

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: zoo/game_2048/config/stochastic_muzero_2048_config.py [hunk 15]
```diff
@@ -0,0 +1,95 @@
+from easydict import EasyDict
```
Comment (PaParaZz1) [inline zoo/game_2048/config/stochastic_muzero_2048_config.py:1]: simplify  config

File: lzero/mcts/tests/tictactoe_muzero_bot_mode_config_test.py [hunk 16]
```diff
@@ -52,6 +52,8 @@
         ),
```
Comment (PaParaZz1) [inline lzero/mcts/tests/tictactoe_muzero_bot_mode_config_test.py:52]: all the tests shoule be renamed to `test_xxx.py`

File: lzero/policy/stochastic_muzero.py [hunk 17]
```diff
@@ -0,0 +1,826 @@
...
+        else:
+            raise ValueError("model type {} is not supported".format(self._cfg.model.model_type))
+
+    def _init_learn(self) -> None:
+        """
+        Overview:
+            Learn mode init method. Called by ``self.__init__``. Ininitialize the learn model, optimizer and MCTS utils.
+        """
+        assert self._cfg.optim_type in ['SGD', 'Adam'], self._cfg.optim_type
+        # NOTE: in board_gmaes, for fixed lr 0.003, 'Adam' is better than 'SGD'.
```
Comment (PaParaZz1) [inline lzero/policy/stochastic_muzero.py:220]: games typo

File: lzero/policy/stochastic_muzero.py [hunk 18]
```diff
@@ -0,0 +1,826 @@
...
+        """
+        Overview:
+            Collect mode init method. Called by ``self.__init__``. Ininitialize the collect model and MCTS utils.
+        """
+        self._collect_model = self._model
+        if self._cfg.mcts_ctree:
+            self._mcts_collect = MCTSCtree(self._cfg)
+        else:
+            self._mcts_collect = MCTSPtree(self._cfg)
+        self.collect_mcts_temperature = 1
```
Comment (PaParaZz1) [inline lzero/policy/stochastic_muzero.py:583]: internal variables should start from underline

File: lzero/policy/stochastic_muzero.py [hunk 19]
```diff
@@ -0,0 +1,826 @@
...
+        value_loss = cross_entropy_loss(value, target_value_categorical[:, 0])
+
+        reward_loss = torch.zeros(self._cfg.batch_size, device=self._cfg.device)
+        consistency_loss = torch.zeros(self._cfg.batch_size, device=self._cfg.device)
+
+        afterstate_policy_loss = torch.zeros(self._cfg.batch_size, device=self._cfg.device)
+        afterstate_value_loss = torch.zeros(self._cfg.batch_size, device=self._cfg.device)
+        commitment_loss = torch.zeros(self._cfg.batch_size, device=self._cfg.device)
+
+        gradient_scale = 1 / self._cfg.num_unroll_steps
```
Comment (PaParaZz1) [inline lzero/policy/stochastic_muzero.py:368]: repeat

File: zoo/game_2048/envs/game_2048_env.py [hunk 20]
```diff
@@ -0,0 +1,757 @@
...
+      - Rendering:
+          The 'render' method can be used to visualize the current state of the game. It supports two rendering modes:
+              - 'human': Renders the game in a text-based format in the console.
+              - 'rgb_array_render': Renders the game as an RGB image.
+          Note: The rendering mode is set to 'human' by default.
+      """
+
+    # The default_config for game 2048 env.
+    config = dict(
+        env_name="game_2048",
```
Comment (PaParaZz1) [inline zoo/game_2048/envs/game_2048_env.py:88]: add per row comments for config fields

File: lzero/model/stochastic_muzero_model.py [hunk 21]
```diff
@@ -0,0 +1,921 @@
...
+    """
+
+    @staticmethod
+    def forward(ctx, input):
+        """
+        Forward method for the one-hot argmax function. This method transforms the input
+        tensor into a one-hot tensor.
+
+        Arguments:
+            ctx (context): A context object that can be used to stash information for
```
Comment (PaParaZz1) [inline lzero/model/stochastic_muzero_model.py:898]: polish comments according to our formats

**Review by PaParaZz1 — APPROVED** [review]

**Pull request merged**